Debug build with gcc fails

pgorszkowski-igalia opened the pull request:
#### fbdc29ad3c83a3384002f0fbe27dfcf4c2763ee7
<pre>
Debug build with gcc fails
<a href="https://bugs.webkit.org/show_bug.cgi?id=258551">https://bugs.webkit.org/show_bug.cgi?id=258551</a>

Reviewed by Youenn Fablet.

Static cast of WebCore::ServiceWorkerState solves a problem with debug
build with gcc. It happens only in debug build because in release build
the RELEASE_ASSERT_WITH_MESSAGE expands to assert without message.

* Source/WebCore/workers/service/server/SWServerWorker.cpp:
(WebCore::SWServerWorker::didFinishInstall):
(WebCore::SWServerWorker::didFinishActivation):

Canonical link: <a href="https://commits.webkit.org/265546@main">https://commits.webkit.org/265546@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b26ddb6a295e09233f1d358a211b6126d6f81f6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11200 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11410 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11741 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12851 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10680 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11221 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13789 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11395 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13596 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11361 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12245 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9463 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13256 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9539 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10139 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17336 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10609 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10294 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13508 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10721 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8807 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9891 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2681 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14165 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10574 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/workers/service/server/SWServerWorker.cpp:
```diff
@@ -201,7 +201,7 @@ void SWServerWorker::didFinishInstall(const std::optional<ServiceWorkerJobDataId
 
     ASSERT(m_server);
-    RELEASE_ASSERT_WITH_MESSAGE(state == ServiceWorkerState::Installing, "State is %hhu", state);
+    RELEASE_ASSERT_WITH_MESSAGE(state == ServiceWorkerState::Installing, "State is %hhu", static_cast<int>(state));
```
Let's cast it to uin8_t since state is enum uint8_t.
RELEASE_LOG though is able to use %hhu, do you know why this works there but not here for you?

pgorszkowski-igalia replied:
It was strange for me too that it works in case of RELEASE_LOG and fails in case RELEASE_ASSERT_WITH_MESSAGE.

The difference in expanding this two macros are:

```
#define RELEASE_LOG(channel, ...) SD_JOURNAL_SEND(channel, LOG_NOTICE, __FILE__, _STRINGIFY(__LINE__), __func__, __VA_ARGS__)

// Expands to
do {
    if (LogServiceWorker.state != WTFLogChannelState ::Off)
        sd_journal_send_with_location("CODE_FILE="
                                      "/home/pgorszkowski/WebKit/Source/WebCore/workers/service/server/SWServerWorker.cpp",
            "CODE_LINE="
            "204",
            __func__, "WEBKIT_SUBSYSTEM=%s", LogServiceWorker.subsystem, "WEBKIT_CHANNEL=%s", LogServiceWorker.name, "PRIORITY=%i", 5, "MESSAGE="
                                                                                                                                       "State is %hhu",
            state, nullptr);
} while (0)
```

```
#define RELEASE_ASSERT_WITH_MESSAGE(assertion, ...) ASSERT_WITH_MESSAGE(assertion, __VA_ARGS__)

// Expands to
do {
    if (!(state == ServiceWorkerState ::Installing)) {
        WTFReportAssertionFailureWithMessage("/home/pgorszkowski/WebKit/Source/WebCore/workers/service/server/SWServerWorker.cpp", 206, __PRETTY_FUNCTION__, "state == ServiceWorkerState::Installing", "State is %hhu", state);
        WTFCrash();
    }
} while (0)
```

I could not find the reason of that. 

I checked and in case LOG with %hhu it also fails in debug.

```
#define LOG(channel, ...)                                     \
    do {                                                      \
        if (LOG_CHANNEL(channel).state != logChannelStateOff) \
            WTFLog(&LOG_CHANNEL(channel), __VA_ARGS__);       \
    } while (0)

// Expands to
do {
    if (LogServiceWorker.state != logChannelStateOff)
        WTFLog(&LogServiceWorker, "State is %hhu", state);
} while (0)
```

pgorszkowski-igalia committed:
Debug build with gcc fails
https://bugs.webkit.org/show_bug.cgi?id=258551

Reviewed by Youenn Fablet.

Static cast of WebCore::ServiceWorkerState solves a problem with debug
build with gcc. It happens only in debug build because in release build
the RELEASE_ASSERT_WITH_MESSAGE expands to assert without message.

* Source/WebCore/workers/service/server/SWServerWorker.cpp:
(WebCore::SWServerWorker::didFinishInstall):
(WebCore::SWServerWorker::didFinishActivation):

Canonical link: https://commits.webkit.org/265546@main
